File: src/jvmTest/kotlin/com/pricecompare/parity/SharedVectorParityTest.kt
package com.pricecompare.parity

import com.pricecompare.model.ComparisonUnit
import com.pricecompare.model.Decimal
import com.pricecompare.model.Money
import com.pricecompare.model.Offer
import com.pricecompare.model.PurchaseContext
import com.pricecompare.model.Quantity
import com.pricecompare.model.TaxMode
import com.pricecompare.pricing.PriceCalculator
import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals

class SharedVectorParityTest {
    @Test
    fun sharedVectorsMatchKotlinEngine() {
        val cases = Json.parseToJsonElement(
            File("test_vectors/price_breakdown_cases.json").readText()
        ).jsonArray

        for (element in cases) {
            val case = element.jsonObject
            val name = case.string("name")
            val offer = Offer(
                productName = name,
                displayedPrice = Money.of(case.string("displayedPrice")),
                taxMode = when (case.string("taxMode")) {
                    "included" -> TaxMode.TAX_INCLUDED
                    "excluded" -> TaxMode.TAX_EXCLUDED
                    else -> error("Unknown tax mode in $name")
                },
                taxRate = Decimal(case.string("taxRate")),
                quantity = Quantity(
                    value = Decimal(case.string("quantity")),
                    unit = ComparisonUnit.fromSymbol(case.string("unit"))
                        ?: error("Unknown unit in $name")
                )
            )
            val context = PurchaseContext(
                shippingCost = Money.of(case.string("shippingCost")),
                usedPoints = case.int("usedPoints"),
                earnedPoints = case.int("earnedPoints"),
                pointEvaluationRate = Decimal(case.string("pointEvaluationRate"))
            )

            val result = PriceCalculator.calculate(offer, context)

            assertEquals(
                case.string("expectedPayableNow"),
                result.payableNow?.amount?.toPlainString(),
                "$name payableNow"
            )
            assertEquals(
                case.string("expectedEffectiveCost"),
                result.effectiveCost?.amount?.toPlainString(),
                "$name effectiveCost"
            )
            assertEquals(
                case.string("expectedUnitPrice"),
                result.unitPrice?.toPlainString(),
                "$name unitPrice"
            )
        }
    }

    private fun JsonObject.string(key: String): String =
        getValue(key).jsonPrimitive.content

    private fun JsonObject.int(key: String): Int =
        getValue(key).jsonPrimitive.content.toInt()
}

File: flutter_app/test/engine/shared_vector_parity_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:price_compare_flutter/engine/decimal.dart';
import 'package:price_compare_flutter/engine/money.dart';
import 'package:price_compare_flutter/engine/offer.dart';
import 'package:price_compare_flutter/engine/price_calculator.dart';
import 'package:price_compare_flutter/engine/purchase_context.dart';
import 'package:price_compare_flutter/engine/quantity.dart';
import 'package:price_compare_flutter/engine/tax_mode.dart';
import 'package:price_compare_flutter/engine/unit.dart';

void main() {
  final cases = _loadCases();

  for (final testCase in cases) {
    test('shared vector: ${testCase['name']}', () {
      final offer = Offer(
        productName: testCase['name']! as String,
        displayedPrice: Money.of(testCase['displayedPrice']! as String),
        taxMode: switch (testCase['taxMode']) {
          'included' => TaxMode.taxIncluded,
          'excluded' => TaxMode.taxExcluded,
          final value => throw FormatException('Unknown tax mode: $value'),
        },
        taxRate: Decimal(testCase['taxRate']! as String),
        quantity: Quantity(
          value: Decimal(testCase['quantity']! as String),
          unit: _unit(testCase['unit']! as String),
        ),
      );
      final context = PurchaseContext(
        shippingCost: Money.of(testCase['shippingCost']! as String),
        usedPoints: testCase['usedPoints']! as int,
        earnedPoints: testCase['earnedPoints']! as int,
        pointEvaluationRate: Decimal(
          testCase['pointEvaluationRate']! as String,
        ),
      );

      final result = PriceCalculator.calculate(offer, context);

      expect(
        result.payableNow.amount.toPlainString(),
        testCase['expectedPayableNow'],
      );
      expect(
        result.effectiveCost.amount.toPlainString(),
        testCase['expectedEffectiveCost'],
      );
      expect(result.unitPrice?.toPlainString(), testCase['expectedUnitPrice']);
    });
  }
}

List<Map<String, Object?>> _loadCases() {
  final raw = File(
    '../test_vectors/price_breakdown_cases.json',
  ).readAsStringSync();
  final decoded = jsonDecode(raw) as List<dynamic>;
  return decoded
      .map((entry) => Map<String, Object?>.from(entry as Map))
      .toList(growable: false);
}

ComparisonUnit _unit(String symbol) => switch (symbol) {
      'g' => ComparisonUnit.gram,
      'kg' => ComparisonUnit.kilogram,
      'ml' => ComparisonUnit.milliliter,
      'L' => ComparisonUnit.liter,
      '個' => ComparisonUnit.count,
      _ => throw FormatException('Unknown comparison unit: $symbol'),
    };
